### PR TITLE
[samsungtv] UPnP devices discovery refactored

### PR DESCRIPTION
**Pull request by kaikreuzer**
Ported https://github.com/openhab/openhab-addons/pull/7499 to openHAB 3.
Fixes https://github.com/openhab/openhab-addons/issues/2916
Fixes https://github.com/openhab/openhab-addons/issues/8419
Fixes https://github.com/openhab/openhab-addons/issues/9495

### CODE REVIEW

**Comment by morph166955** [comment]
@kai when you get this done, happy to test if you want to post a jar.

**Comment by kaikreuzer** [comment]
I think this looks good, so ready for review by some other maintainer.

**Comment by morph166955** [comment]
I compiled and loaded the jar for this to see if it resolves the thread issues from the original report. So far no issues.

**Review by cpmeister — APPROVED** [review]
LGTM

**Comment by jaywiseman1971** [comment]
Will there be a fix for the Samsung 2.5.x version if this works?

Best, Jay

**Comment by kaikreuzer** [comment]
@jaywiseman1971 Just checked, it is not as simple as a cherry-pick to the 2.5.x branch, so it'd probably be a bit more work&testing.
Let's first see it working for 3.x and then we can see. Maybe it would also be a incentive for you to migrate ;-) .

**Comment by kaikreuzer** [comment]
@paulianttila Are you around for a quick second review or shall we merge it right away?

File: bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java
Comment (paulianttila) [inline bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:152]: There might be a small chance to miss device(s) if new is registered between getRemoteDevices loop and addListener.

File: bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java
Comment (paulianttila) [inline bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:367]: Some of the Samsung TV models have a faulty UPnP implementation as they don't publish SSDP alive message as they have promised and jUPnP library might make a decision to remove the device. If device is a MediaRenderer device, shutdown now remove all samsung tv related UPnP services and they might not come back as remoteDeviceAdded is not called by the UPnP library for those removed services. Earlier checkAndCreateServices was called which created all related services.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:367]: What's your suggestion on what to do here then? Instead of doing shutdown/offline call `checkCreateManualConnection`? Or simply ignore that call and assume that services are still available?

Why is it different to the previous `thingRemoved` being called by the UPnPDiscoveryService when the UPnP service disappears from JUPnP?

File: bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java]: Addressed.

File: bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java
Comment (paulianttila) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: When remoteDeviceAdded is called for MediaRenderer devices, binding could try to create all related services like it was done earlier by checkAndCreateServices.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: Why should it try to create services, if all are still there?

And please for me to understand: Why is it different to the previous thingRemoved being called by the UPnPDiscoveryService when the UPnP service disappears from JUPnP?

Comment (paulianttila) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: ThingRemoved is not the issue, but the thingDiscovered. 

1. Samsung TV binding is up and running (MediaRenderer service and e.g. MainTV service are up)
2. MediaRenderer device disappears, so JUPnP library calls thingRemoved for MediaRenderer device.
3. During thingRemoved method, binding remove **all** services (MediaRenderer and MainTV) by shutdown method.
4. MediaRenderer device is discovered again and remoteDeviceAdded is called.
5. Binding creates MediaRenderer service.
6. MainTV service is missing

Comment (paulianttila) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: I'm not using Samsung TV binding my self at the moment as I have separated VLAN's and Samsung TV doesn't allow communication from another network. Some users have been stated  that 2.4 version have been working very reliable.

https://github.com/openhab/openhab2-addons/blob/2.4.0/addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java

2.4 binding implemented both DiscoveryListener and RegistryListener.

thingDiscovered -> create all available services
thingRemoved -> shutdown all services
remoteDeviceAdded -> create discovered service

Comment (jaywiseman1971) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: > users have been stated that 2.4 version have been working very reliable

Yup, 2.4 RC1 is very stable.  I tried to work with folks on this earlier this year to get 2.5 version to work which introduces new Samsung TV models.   I even put up a bounty to fix this . . .

Best, Jay

Comment (kaikreuzer) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: > ThingRemoved is not the issue, but the thingDiscovered.

Sorry, but I still do not understand. Do you refer to this PR or about an issue that already exists in the current 2.5/3.0 code base?

Note that this PR only tries to fix the issue that the binding currently causes deadlocks that affect other bindings and the core runtime as a whole.
If there are issues to solve with unreliable TV detection due to the handling of SSDP messages (or their non-existence due to buggy TVs), please let us discuss those in a different issue.

Comment (paulianttila) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: >Do you refer to this PR

Yes

>Note that this PR only tries to fix the issue that the binding currently causes deadlocks that affect other bindings and the core runtime as a whole.

Yes I know, but this PR also change the behaviour of the binding, which I think might cause some issues.

Let's try one more time

On master, thingDiscovered method calls checkAndCreateServices (line 362), which will go throw all existing UPnP devices and create all necessary Samsung tv services.

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L333-L364

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L246-L254

All Samsung tv related services are removed in thingRemoved method when service is MediaRenderer device

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L367-L373

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L183-L189


In this PR, thingDiscovered and thingRemoved are removed and it fully relie to UPnP events (remoteDeviceAdded and remoteDeviceRemoved). When remoteDeviceRemoved is done for MediaRenderer device, it shutdown all existing services, but remoteDeviceAdded does only create service which was found, so following could happen

1. Samsung TV binding is up and running (MediaRenderer service and e.g. MainTV service are up)
2. MediaRenderer device disappears, so JUPnP library calls remoteDeviceRemoved for MediaRenderer device.
3. remoteDeviceRemoved method remove all services (MediaRenderer and MainTV) by shutdown method.
4. MediaRenderer device is discovered again and remoteDeviceAdded method is called.
5. Binding creates MediaRenderer service.
6. MainTV service is still missing, so only half of the binding features are working 

So, when MediaRenderer device is added, binding should check and create all Samsung TV related services (then it match equally the master branch functionality) or remoteDeviceAdded should not shutdown all services, but only service which was removed.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: > Let's try one more time

Thanks a lot for your patience with me. 🙏 

> On master, thingDiscovered method calls checkAndCreateServices (line 362), which will go throw all existing UPnP devices and create all necessary Samsung tv services.

So if I understand correctly, this is exactly the feature you are missing now with that PR, right? My confusion might come from the fact that we are discussing this in the `remoteDeviceRemoved` method.

> All Samsung tv related services are removed in thingRemoved method 

Yes, that's what I [referred to above](https://github.com/openhab/openhab-addons/pull/9525#discussion_r549727193) - it is the very same as what happens in the code of this PR, that's why I couldn't see any problem with the `remoteDeviceRemoved` method.

> So, when MediaRenderer device is added, binding should check and create all Samsung TV related services 

That should be easy to do in the `remoteDeviceAdded` method then. But it still is not clear to me why this solves the problem you mentioned initially: "they don't publish SSDP alive message". If there are no SSDP messages, we will only see a call to `remoteDeviceRemoved`, which falsely takes the Thing OFFLINE, but will never see a call to `remoteDeviceAdded`, which would bring it back ONLINE again. That's why I am confused on what should be done.

**Comment by morph166955** [comment]
I'm not sure if this is relevant to this bug/PR, and I'll open a separate issue if so.  On my current setup, the websocket based TVs keep their thing status as UNINITIALIZED on OH startup until the TV is actually turned on for "the first time".  Then it goes ONLINE/OFFLINE.  The older TVs go right to OFFLINE.  2.4.0RC1 used to have the websocket based TVs go to OFFLINE.

File: bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java
Comment (paulianttila) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: I didn't say that TV don't publish SSDP alive messages at all, but not like they have promised (max-age for advertisement), so UPnP library think that service is offline and then suddenly TV send the advertisement. Behaviour seems to be also service specific and that's why there are many questions on the forum like yesterday day everything worked fine and today I can change the volume but not the channel. Which of course is weird from user perspective as they don't understand that TV contains multiple UPnP services and each service support different features.

Comment (paulianttila) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:370]: Few years ago, I tried to find solution for this and found following from Cling page, but I never found time to test it. It will impact to all UPnP devices and binding, but maybe something clever could implemented to OH side or JUPnP.

http://4thline.org/projects/cling/core/manual/cling-core-manual.xhtml#section.RemoteMaxAge

Anyhow, idea was to increase the max age or disable it and then manually but thing offline state if several consecutive queries fails (TV power is OFF).

Comment (kaikreuzer) [reply bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java:363]: Ok, thanks again @paulianttila.
I have now reverted everything from #7499 that was not directly related to replacing thing discovery methods by upnp device discovery methods.
This leaves the code hopefully working exactly the same way as before, while still not causing deadlocks anymore caused by the discovery listener registration.

**Comment by kaikreuzer** [comment]
@morph166955 Please let us cover that in a different issue (unless already fixed by this PR as a side-effect...).

**Comment by morph166955** [comment]
No issue here.  I wasn't sure if it was relevant with everything you were doing, only reason I brought it up.  I'll open a separate bug once this gets committed to make sure it's still relevant with your changes.

**Comment by morph166955** [comment]
@kaikreuzer I think there may be an issue with the current version.  This morning my kids accidentally turned on the TV, I turned it off very quickly after (a few seconds).  Since then, the thing for that TV has been stuck in an online state.  I've turned it on/off about a dozen times, and it will not budge.  The network binding is reporting that the pings are stopping when the TV is off so I'm not sure how/why upnp is not releasing the TV back to an offline state.  I did a stack trace, attached below.  

[samsung-stuck.txt](https://github.com/openhab/openhab-addons/files/5760798/samsung-stuck.txt)

**Pull request merged**
